### PR TITLE
Update infrastructure

### PR DESCRIPTION
**Pull request by timmens**
### What problem do you want to solve?

Match the code base infrastructure with that of [estimagic](http://github.com/opensourceeconomics/estimagic).

### Changes

- Replace testing with tox with pure pytest testing
- Update main workflow file
- Add publish to pypi workflow 
- Update issue and pull request templates
- Update .gitignore
- Update pre-commit and linting files
- Update LICENSE
- Replace RST files with Markdown files (except for documentation files)
- Delete outdated documentation
- Use ruff blacklisting (depart from estimagic here)

### Todo

- [x] Target the right branch and pick an appropriate title.
- [x] Put `Closes #XXXX` in the first PR comment to auto-close the relevant issue once
  the PR is accepted. This is not applicable if there is no corresponding issue.
- [x] Update CHANGES.md

### CODE REVIEW

**Review by ChristianZimpelmann — COMMENTED** [review]
I had a look over it and made a few small comments.

Also:

- Don't you think the documentation could still be of use? It seems to be quite general.. but haven't had a closer look at it.

- I cannot say much about some parts of the setup (e.g. tox vs pytest, ruff blacklisting)

File: LICENSE [hunk 1]
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2021 IZA - Institute of Labor Economics
+Copyright (c) 2023 Hans-Martin von Gaudecker
```
[the remark this answers is not in the document]
Comment (hmgaudecker) [reply LICENSE:3]: > Do we need to name a person here? 

Definitely not me!

[the remark this answers is not in the document]
Comment (timmens) [reply LICENSE:3]: What about OpenSourceEconomics?

[the remark this answers is not in the document]
Comment (hmgaudecker) [reply LICENSE:3]: Fine. We should probably add a few key names and then "et al." as OSE is not registered in any way.

**Comment by timmens** [comment]
Thank you @ChristianZimpelmann 

> * Don't you think the documentation could still be of use? It seems to be quite general.. but haven't had a closer look at it.

Janos recommended deleting it. The docs were comprised mostly of _design notes_, which are rather developer documentation. I think this developer documentation is very important, but I want to use a different format ---something not on GitHub. Some of Janos' notes would be integrated into this new format.

> * I cannot say much about some parts of the setup (e.g. tox vs pytest, ruff blacklisting)

Maybe @hmgaudecker can comment on this.

**Comment by hmgaudecker** [comment]
Happy to do so, but unlikely to happen before our meeting.

File: src/lcm/discrete_emax.py [hunk 2]
```diff
@@ -211,13 +212,10 @@ def _determine_discrete_choice_axes(variable_info):
     """
     has_sparse = variable_info["is_sparse"].any()
     dense_vars = variable_info.query(
-        "is_dense & ~(is_choice & is_continuous)"
+        "is_dense & ~(is_choice & is_continuous)",
```
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply src/lcm/discrete_emax.py:215]: I see. Seems strange to me as there is only one argument and I have never used more than one argument calling `pd.DataFrame.query`. But I should get used to it :)

**Comment by ChristianZimpelmann** [comment]
> The docs were comprised mostly of design notes, which are rather developer documentation. I think this developer documentation is very important, but I want to use a different format ---something not on GitHub. Some of Janos' notes would be integrated into this new format.

That makes sense! Just wanted to make sure that the design notes are not lost.

File: pyproject.toml [hunk 3]
```diff
@@ -4,3 +4,112 @@ requires = ["setuptools>=45", "wheel", "setuptools_scm[toml]>=6.0"]
...
+    # boolean default value in function definition
+    "FBT002",
+
+    # Too many arguments to function call
+    "PLR0913",
+
+    # magic value used in comparison
+    "PLR2004",
+
+    # unnecessary variable assignment before return
```
Comment (hmgaudecker) [inline pyproject.toml:56]: Would like to keep it. Trivial to change and doing this

```py
out = whatever(x, y, z)
return out
```

just adds a complication to thinking about code relative to

```py
return whatever(x, y, z)
```

IMHO.

File: pyproject.toml [hunk 4]
```diff
@@ -4,3 +4,112 @@ requires = ["setuptools>=45", "wheel", "setuptools_scm[toml]>=6.0"]
...
+    "ignore:distutils Version classes are deprecated. Use packaging.version instead",
+    "ignore:Standard matrix inversion failed due to LinAlgError",
+    "ignore:Tranquilo is extremely experimental",
+    "ignore:delta_grad == 0.0",
+    "ignore:Widget._active_widgets is deprecated",
+    "ignore:Widget._widget_types is deprecated",
+    "ignore:Widget.widget_types is deprecated",
+    "ignore:Widget.widgets is deprecated",
+]
+addopts = ["--doctest-modules"]
```
Comment (hmgaudecker) [inline pyproject.toml:103]: I don't think we'll ever have doctests.

File: pyproject.toml [hunk 5]
```diff
@@ -4,3 +4,112 @@ requires = ["setuptools>=45", "wheel", "setuptools_scm[toml]>=6.0"]
...
+markers = [
+    "wip: Tests that are work-in-progress.",
+    "slow: Tests that take a long time to run and are skipped in continuous integration.",
+    "jax: Tests that require jax to be installed and are skipped on non-Linux systems.",
```
Comment (hmgaudecker) [inline pyproject.toml:107]: remove
```suggestion
```

File: setup.cfg [hunk 6]
```diff
@@ -1,37 +1,39 @@
...
 
 [options]
 packages = find:
 install_requires =
+    dags
     estimagic
-python_requires = >=3.6.1
+    jax
+    numpy
+    pandas
```
Comment (hmgaudecker) [inline setup.cfg:35]: Shouldn't this include everything from testenv? dags, networkx, pybaum, ...

File: setup.cfg [hunk 7]
```diff
@@ -43,3 +45,7 @@ where = src
 [options.entry_points]
 console_scripts =
```
Comment (hmgaudecker) [inline setup.cfg:46]: Do we have a cli?

File: src/lcm/distributions.py [hunk 8]
```diff
@@ -3,15 +3,15 @@
 
 
 def lognormal(sd, size):
-    return np.random.lognormal(0, sd, size)
+    return np.random.lognormal(0, sd, size)  # noqa: NPY002
```
Comment (hmgaudecker) [inline src/lcm/distributions.py:6]: Pls open an issue to search & remove NPY002 https://numpy.org/doc/stable/reference/random/legacy.html#legacy

Will stop working at some point, so better make the move soon.

File: src/lcm/model_functions.py [hunk 9]
```diff
@@ -43,7 +43,7 @@ def get_utility_and_feasibility_function(
     """
     if is_last_period:
         # vf_arr is there just so the final function takes a vf_arr, without using it.
```
Comment (hmgaudecker) [inline src/lcm/model_functions.py:45]:
```suggestion
        # vf_arr is there just so the final period function takes a vf_arr, without using it.
```

**Review by hmgaudecker — COMMENTED** [review]
Excellent, thanks! I do have many comments, but almost all of them should be very quick and none of them requires any discussion; if you want to leave some of the ruff things as they are, just make a quick note, please.

File: src/lcm/distributions.py (hunk 8 above)
Comment (timmens) [reply src/lcm/distributions.py:6]: Done in #6

File: .envs/testenv.yml
[the remark this answers is not in the document]
Comment (timmens) [reply .envs/testenv.yml:1]: I prefer having two files to minimize the chance of forgetting to add packages to `setup.cfg` or running the test suite in the development environment.

File: .gitignore [hunk 10]
```diff
@@ -1,24 +1,131 @@
...
-_build
-__pycache__
-_generated
 
-*.egg-info
-.eggs
+# Rope project settings
+.ropeproject
 
-.*
```
[the remark this answers is not in the document]
Comment (timmens) [reply .gitignore:17]: The .gitignore is copied verbatim from estimagic, and we did not have problems there. If you prefer to whitelist hidden files we can do that though.

[the remark this answers is not in the document]
Comment (hmgaudecker) [reply .gitignore:17]: No worries!

File: setup.cfg (hunk 6 above)
Comment (timmens) [reply setup.cfg:35]: I'd say no. The testing process requires other packages, e.g., `scipy`, compared to the package itself. I've added comments to the environment files that should make this clear.

Comment (hmgaudecker) [reply setup.cfg:35]: Okay, maybe not everything, but anything that is imported in lcm proper, right?

Comment (timmens) [reply setup.cfg:35]: Yes, exactly. I think this should be the case now after my latest updates.

**Comment by timmens** [comment]
Leaving docformatter at v1.5.1 instead of v.1.6.0 because of incompatibilities with black.

File: pyproject.toml [hunk 11]
```diff
@@ -4,3 +4,112 @@ requires = ["setuptools>=45", "wheel", "setuptools_scm[toml]>=6.0"]
...
+    # missing return type annotation for public function
+    "ANN201",
+
+    # missing return type annotation for private function
+    "ANN202",
+
+    # `zip()` without an explicit `strict` parameter
+    "B905",
+
+    # Missing docstring in public module
```
[the remark this answers is not in the document]
Comment (timmens) [reply pyproject.toml:32]: Yes, I fully agree. I've opened an issue.

File: pyproject.toml [hunk 12]
```diff
@@ -4,3 +4,112 @@ requires = ["setuptools>=45", "wheel", "setuptools_scm[toml]>=6.0"]
...
+    # missing docstring in public package
+    "D104",
+
+    # exception must not use a string literal
+    "EM101",
+
+    # exception must not use an f-string literal
+    "EM102",
+
+    # boolean default value in function definition
```
[the remark this answers is not in the document]
Comment (timmens) [reply pyproject.toml:47]: I've allowed the rule for now. But I haven't ultimately decided if I like it or not. Will speak to Janos and Tobi about this.

File: pyproject.toml [hunk 13]
```diff
@@ -4,3 +4,112 @@ requires = ["setuptools>=45", "wheel", "setuptools_scm[toml]>=6.0"]
...
+    "FBT002",
+
+    # Too many arguments to function call
+    "PLR0913",
+
+    # magic value used in comparison
```
[the remark this answers is not in the document]
Comment (timmens) [reply pyproject.toml:53]: I'm ignoring it now for test files but not for the package.

**Comment by hmgaudecker** [comment]
> Leaving docformatter at v1.5.1 instead of v.1.6.0 because of incompatibilities with black.

Let's just delete it, ruff should take care of it eventually.

**Review by hmgaudecker — APPROVED** [review]
Just for Github to remember I browsed through changes up to this point.

**Comment by hmgaudecker** [comment]
(I actually did not mean to approve, else the message would not have made sense, but I might as well have. So just merge as you see fit)

**Thanks!**

**Pull request merged**